packages/styled-components/src/models/StyleSheetManager.js

```diff
@@ -15,41 +15,26 @@ type Props = {
 
 export const StyleSheetContext: Context<StyleSheet | void> = React.createContext();
 export const StyleSheetConsumer = StyleSheetContext.Consumer;
+export const StylisContext: Context<Stringifier | void> = React.createContext();
+export const StylisConsumer = StylisContext.Consumer;
+
 export const masterSheet: StyleSheet = new StyleSheet();
+export const masterStylis: Stringifier = createStylisInstance();
 
 export function useStyleSheet(): StyleSheet {
   return useContext(StyleSheetContext) || masterSheet;
 }
 
-export default function StyleSheetManager(props: Props) {
-  /**
-   * freeze the stylis modification props on initial mount since they rely on
-   * reference equality for the useMemo dependencies array and devs will
-   * likely not store the reference themselves to avoid this issue
-   */
-  const [{ disableVendorPrefixes, stylisPlugins }] = useState({
-    disableVendorPrefixes: props.disableVendorPrefixes,
-    stylisPlugins: props.stylisPlugins,
-  });
-
-  if (process.env.NODE_ENV !== 'production') {
-    if (!shallowequal(disableVendorPrefixes, props.disableVendorPrefixes)) {
-      // eslint-disable-next-line no-console
-      console.warn(
-        'disableVendorPrefixes is frozen on initial mount of StyleSheetManager. Changing this prop dynamically will have no effect.'
-      );
-    }
+export function useStylis(): Stringifier {
+  return useContext(StylisContext) || masterStylis;
+}
 
-    if (!shallowequal(stylisPlugins, props.stylisPlugins)) {
-      // eslint-disable-next-line no-console
-      console.warn(
-        'stylisPlugins are frozen on initial mount of StyleSheetManager. Changing this prop dynamically will have no effect.'
-      );
-    }
-  }
+export default function StyleSheetManager(props: Props) {
+  const [plugins, setPlugins] = useState(props.stylisPlugins);
+  const contextStyleSheet = useStyleSheet();
 
   const styleSheet = useMemo(() => {
-    let sheet = masterSheet;
+    let sheet = contextStyleSheet;
```

Comment:
This fixes some of the SSR problems around nested StyleSheetManager (since collectStyles adds its own)